File: implementations/rust/src/bin/benchmark.rs
use b57::{encoded_length, decoded_length, encode, decode, id57_generate_default};
use std::time::Instant;

fn main() {
    let iterations = 100_000;
    let mut data = [0u8; 32];
    let mut encoded_strs = Vec::with_capacity(iterations);

    let start = Instant::now();
    for i in 0..iterations {
        data[0] = (i & 0xFF) as u8;
        data[1] = ((i >> 8) & 0xFF) as u8;
        encoded_strs.push(encode(&data));
    }
    let encode_ms = start.elapsed().as_millis();

    let start = Instant::now();
    for i in 0..iterations {
        let _ = decode(&encoded_strs[i]).unwrap();
    }
    let decode_ms = start.elapsed().as_millis();

    let start = Instant::now();
    for i in 0..iterations {
        data[0] = (i & 0xFF) as u8;
        data[1] = ((i >> 8) & 0xFF) as u8;
        let _ = id57_generate_default(&data);
    }
    let id_ms = start.elapsed().as_millis();

    println!(
        "{{\"language\": \"Rust\", \"encode_ms\": {}, \"decode_ms\": {}, \"id57_ms\": {}, \"iterations\": {}}}",
        encode_ms, decode_ms, id_ms, iterations
    );
}

File: implementations/dart/bin/benchmark.dart
import 'dart:convert';
import 'package:f57/b57.dart';

void main() {
  const iterations = 100000;
  final data = List<int>.filled(32, 0);
  final encodedStrs = List<String>.filled(iterations, "");

  var start = DateTime.now();
  for (var i = 0; i < iterations; i++) {
    data[0] = i & 0xFF;
    data[1] = (i >> 8) & 0xFF;
    encodedStrs[i] = encode(data);
  }
  final encodeMs = DateTime.now().difference(start).inMilliseconds;

  start = DateTime.now();
  for (var i = 0; i < iterations; i++) {
    decode(encodedStrs[i]);
  }
  final decodeMs = DateTime.now().difference(start).inMilliseconds;

  start = DateTime.now();
  for (var i = 0; i < iterations; i++) {
    data[0] = i & 0xFF;
    data[1] = (i >> 8) & 0xFF;
    id57GenerateDefault(data);
  }
  final idMs = DateTime.now().difference(start).inMilliseconds;

  print(jsonEncode({
    "language": "Dart",
    "encode_ms": encodeMs,
    "decode_ms": decodeMs,
    "id57_ms": idMs,
    "iterations": iterations
  }));
}
